workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs

```diff
@@ -10,77 +10,137 @@ namespace Improbable.Gdk.Core
 {
     public class Worker : IDisposable
     {
-        public delegate void WorkerEventHandler(Worker worker);
-
-        public static event WorkerEventHandler OnConnect;
-        public static event WorkerEventHandler OnDisconnect;
-
         public readonly Vector3 Origin;
         public readonly string WorkerType;
         public readonly string WorkerId;
-        public readonly Entity WorkerEntity;
         public readonly ILogDispatcher LogDispatcher;
 
         public Connection Connection { get; private set; }
         public World World { get; private set; }
 
+        private readonly WorkerDisconnectCallbackSystem disconnectCallbackSystem;
 
-        internal readonly Dictionary<EntityId, Entity> EntityMapping = new Dictionary<EntityId, Entity>();
+        public event Action<string> OnDisconnect
+        {
+            add => disconnectCallbackSystem.OnDisconnected += value;
+            remove => disconnectCallbackSystem.OnDisconnected -= value;
+        }
 
-        private Worker(ConnectionConfig config, ILogDispatcher logDispatcher, Vector3 origin)
+        private Worker(string workerType, Connection connection, ILogDispatcher logDispatcher, Vector3 origin)
         {
             Origin = origin;
-            WorkerId = string.IsNullOrEmpty(config.WorkerId)
-                ? $"{config.WorkerType}-{Guid.NewGuid()}"
-                : config.WorkerId;
+            WorkerType = workerType;
+            WorkerId = connection.GetWorkerId();
 
-            Connection = ConnectionUtility.Connect(config, WorkerId);
+            Connection = connection;
             LogDispatcher = logDispatcher;
-            WorkerType = config.WorkerType;
             logDispatcher.Connection = Connection;
 
-            World = new World(WorkerId);
-            World.CreateManager<WorkerSystem>(this);
-            var entityManager = World.GetOrCreateManager<EntityManager>();
-            WorkerEntity = entityManager.CreateEntity(typeof(OnConnected), typeof(WorkerEntityTag));
-            OnConnect?.Invoke(this);
+            World = new World(Connection.GetWorkerId());
+            AddCoreSystems();
+
+            // This isn't a core system, this is for an easy disconnect event
+            disconnectCallbackSystem = World.GetOrCreateManager<WorkerDisconnectCallbackSystem>();
```

Comment:
But we also have the `OnDisconnect event in this class. If a system wants to react to a worker disconnecting, it can just inject the `OnDisconnect` tag.
If a monobehaviour wants to react to it, it will use the event from this class. So I don't see a reason to add events to systems.

Reply:
Then we have to chain together the disconnect events. This class can't know about a disconnection without a system telling it. This is less redundancy than the alternative.